cf/data/data.py

```diff
@@ -9110,37 +9123,57 @@ def equals(
         other_Units = other.Units
         if self_Units != other_Units:
             logger.info(
-                "{}: Different Units ({!r}, {!r}".format(
-                    self.__class__.__name__, self.Units, other.Units
-                )
+                f"{self.__class__.__name__}: Different Units "
+                f"({self.Units!r}, {other.Units!r})"
             )
             return False
 
-        config = self.partition_configuration(readonly=True)
-
-        other.to_memory()
-
-        for partition in self.partitions.matrix.flat:
-            partition.open(config)
-            array0 = partition.array
-            array1 = other[partition.indices].varray
-            partition.close()
+        self_dx = self._get_dask()
+        other_dx = other._get_dask()
+
+        # Now check that corresponding elements are equal within a tolerance.
+        # We assume that all inputs are masked arrays. Note we compare the
+        # data first as this may return False due to different dtype without
+        # having to wait until the compute call.
+        self_is_numeric = _is_numeric_dtype(self_dx)
+        other_is_numeric = _is_numeric_dtype(other_dx)
+        if self_is_numeric and other_is_numeric:
+            print("DATA COMP'ING IS", self_dx.compute(), other_dx.compute())
+            data_comparison = _da_ma_allclose(
+                self_dx,
+                other_dx,
+                masked_equal=False,  # TODODASK: is this correct, or want True?
+                rtol=float(rtol),
+                atol=float(atol),
+            )
+            print("DATA COMP RESULT IS", data_comparison.compute())
+        elif not self_is_numeric and not other_is_numeric:
+            data_comparison = da.all(self_dx == other_dx)
+        else:  # one is numeric and other isn't => not equal (incompat. dtype)
+            logger.info(
+                f"{self.__class__.__name__}: Different data types:"
+                f"{self_dx.dtype} != {other_dx.dtype}"
+            )
+            return False
 
-            if not _numpy_allclose(
-                array0, array1, rtol=float(rtol), atol=float(atol)
-            ):
-                logger.info(
-                    "{0}: Different array values (atol={1}, "
-                    "rtol={2})".format(self.__class__.__name__, atol, rtol)
-                )
+        mask_comparison = da.all(
```

Comment:
I am not convinced it would be more efficient because I would have thought `all` would be most efficient due to its nature, unless I am missing something here? I can profile them to see... will let you know and choose the most efficient one.

Reply:
You're quite right. I withdraw my suggestion. `da.all(da.equal())` is certainly faster for boolean arrays than `da.allclose()`